clang/lib/CIR/CodeGen/CIRGenVTables.cpp

```diff
@@ -414,6 +422,118 @@ CIRGenModule::getVTableLinkage(const CXXRecordDecl *RD) {
+      .mangleCXXVTT(RD, Out);
+  StringRef Name = OutName.str();
+
+  // This will also defer the definition of the VTT.
+  (void)CGM.getCXXABI().getAddrOfVTable(RD, CharUnits());
+
+  VTTBuilder Builder(CGM.getASTContext(), RD, /*GenerateDefinition=*/false);
+
+  auto ArrayType = mlir::cir::ArrayType::get(CGM.getBuilder().getContext(),
+                                             CGM.getBuilder().getUInt8PtrTy(),
+                                             Builder.getVTTComponents().size());
+  auto Align =
+      CGM.getDataLayout().getABITypeAlign(CGM.getBuilder().getUInt8PtrTy());
+  auto VTT = CGM.createOrReplaceCXXRuntimeVariable(
+      CGM.getLoc(RD->getSourceRange()), Name, ArrayType,
+      mlir::cir::GlobalLinkageKind::ExternalLinkage,
+      CharUnits::fromQuantity(Align));
+  CGM.setGVProperties(VTT, RD);
+  return VTT;
+}
+
+/// Emit the definition of the given vtable.
+void CIRGenVTables::buildVTTDefinition(mlir::cir::GlobalOp VTT,
+                                       mlir::cir::GlobalLinkageKind Linkage,
+                                       const CXXRecordDecl *RD) {
+  VTTBuilder Builder(CGM.getASTContext(), RD, /*GenerateDefinition=*/true);
+
+  auto ArrayType = mlir::cir::ArrayType::get(CGM.getBuilder().getContext(),
+                                             CGM.getBuilder().getUInt8PtrTy(),
+                                             Builder.getVTTComponents().size());
+
+  SmallVector<mlir::cir::GlobalOp, 8> VTables;
+  SmallVector<VTableAddressPointsMapTy, 8> VTableAddressPoints;
+  for (const VTTVTable *i = Builder.getVTTVTables().begin(),
+                       *e = Builder.getVTTVTables().end();
+       i != e; ++i) {
+    VTableAddressPoints.push_back(VTableAddressPointsMapTy());
+    VTables.push_back(getAddrOfVTTVTable(*this, CGM, RD, *i, Linkage,
+                                         VTableAddressPoints.back()));
+  }
+
+  SmallVector<mlir::Attribute, 8> VTTComponents;
+  for (const VTTComponent *i = Builder.getVTTComponents().begin(),
+                          *e = Builder.getVTTComponents().end();
+       i != e; ++i) {
+    const VTTVTable &VTTVT = Builder.getVTTVTables()[i->VTableIndex];
+    mlir::cir::GlobalOp VTable = VTables[i->VTableIndex];
+    VTableLayout::AddressPointLocation AddressPoint;
+    if (VTTVT.getBase() == RD) {
+      // Just get the address point for the regular vtable.
+      AddressPoint =
+          getItaniumVTableContext().getVTableLayout(RD).getAddressPoint(
+              i->VTableBase);
+    } else {
+      AddressPoint = VTableAddressPoints[i->VTableIndex].lookup(i->VTableBase);
+      assert(AddressPoint.AddressPointIndex != 0 &&
+             "Did not find ctor vtable address point!");
+    }
+
+    mlir::Attribute Idxs[3] = {
+        CGM.getBuilder().getI32IntegerAttr(0),
+        CGM.getBuilder().getI32IntegerAttr(AddressPoint.VTableIndex),
+        CGM.getBuilder().getI32IntegerAttr(AddressPoint.AddressPointIndex),
```

Comment:
Instead of just passing opaque integers, we should generalize address point as well, since it makes it easier to understand vtables in a higher level. To be more specific, see `VTableAddrPointOp` two input attributes, note they have some pretty printing - it's the same stuff here, we should create a new attribute for an address point, and use it in both `VTableAddrPointOp` and here. Later LLVM codegen can decouple into an actual array.

Reply:
You mean a const struct like attribute for address point? The `Idxs` attribute is needed by the `GlobalViewAttr` below. I'm not sure how a struct attribute works there.

Reply:
I mean a new attribute that wraps address points, it's disconnect from structs (though that's an easy assumption to make because it's a few integers, so it looks like one). `VTableAddrPointOp` has that embedded because it was the only user.

Reply:
The user here is a `GlobalViewAttr` that accepts an index array. I don't see how a new attribute works here unless we wrap the `GlobalViewAttr` with another new attribute as well. Not sure it's worth the complexity. Maybe I misunderstand your point.

Reply:
I have a few ideas. We can tackle it later in #258